transformer_engine/pytorch/attention.py

```diff
@@ -3596,26 +3628,17 @@ def forward(
             batch_start = inference_params.batch_size_offset
             batch_end = batch_start + key_layer.size(1)
             assert batch_end <= inference_key_memory.size(1)
+
             sequence_start = inference_params.sequence_len_offset
             sequence_end = sequence_start + key_layer.size(0)
             assert sequence_end <= inference_key_memory.size(0)
-            # Copy key and values.
-            inference_key_memory[
-                sequence_start:sequence_end, batch_start:batch_end, ...
-            ] = key_layer
-            inference_value_memory[
-                sequence_start:sequence_end, batch_start:batch_end, ...
-            ] = value_layer
-            key_layer = inference_key_memory[:sequence_end, batch_start:batch_end, ...]
-            value_layer = inference_value_memory[
-                :sequence_end, batch_start:batch_end, ...
-            ]
 
             # adjust the key rotary positional embedding
             if rotary_pos_emb is not None:
+                delta = sequence_end - sequence_start
                 q_pos_emb, k_pos_emb = rotary_pos_emb
-                q_pos_emb = q_pos_emb[sequence_start:sequence_end, :, :, :]
-                k_pos_emb = k_pos_emb[:sequence_end, :, :, :]
+                q_pos_emb = q_pos_emb[:delta, :, :, :]
+                k_pos_emb = k_pos_emb[:delta, :, :, :]
                 rotary_pos_emb = (q_pos_emb, k_pos_emb)
```

Comment:
I agree with Tim's suggestion - you need to take the embeddings in the right positions for q and k.

For the change in the execution order between RoPE and KV-cache retrieval -> we were discussing it with @ksivaman and @Oleg-Goncharov and we believe this is fine change to make, since in the end KV-cache is just a cache of the values being passed to attention in order to avoid doing the full sequence length every time. It is an optimization of DPA itself about its inputs so it should not matter what those inputs were before (whether they were actually K and V or something completely different). This also makes RoPE logic faster during inference since we are no longer applying it to the fdull sequence every time.

Reply:
It would be good to understand why your test worked with the current code since that delta change should change the values.